File: bexar_models.R
# Bexar County — Prosecutor Learning Curve Regression Models
# Requires: bexar_prosecutor_panel_1990_2015.parquet (from bexar_prosecutor_panel.R)
# install.packages(c("tidyverse", "arrow", "broom", "fixest"))
#
# Key outcome: DEFERRED
# Key exposure: PROSECUTOR_CASE_N100 × RACE interactions
# The interaction coefficient tests whether the White-Black gap widens with experience.
#
# Main table specs: SpecA (offense controls), SpecB (+ year FE), SpecC (prosecutor FE)
# Robustness: M1 (baseline), M2 (offense FE), M6 (appointed only),
#             M7 (within offense type), M8 (plea-conditional), M9 (straight conviction)
# Output: bexar_model_results.csv

DATA_DIR <- "C:/Users/carolineferguson/Box/Bigelow/Bexar/Data"

library(tidyverse)
library(arrow)
library(broom)
library(fixest)

dp <- read_parquet(file.path(DATA_DIR, "bexar_prosecutor_panel_1990_2015.parquet"))

# ── Analysis frame ────────────────────────────────────────────────────────────

df <- dp |>
  filter(`RACE-LABEL` %in% c("Black", "White", "Latino")) |>
  mutate(
    BLACK        = as.integer(`RACE-LABEL` == "Black"),
    LATINO       = as.integer(`RACE-LABEL` == "Latino"),
    RACE         = fct_relevel(`RACE-LABEL`, "White"),
    OFFENSE_TYPE = fct_relevel(`OFFENSE-CLASS`, "F3"),
    OFFENSE_CATEGORY2 = case_when(
      str_detect(`OFFENSE-DESC`, "POSS CS|POSS W/I DEL CS|POSS W/INT DEL CS|MAN/DEL CS|DEL CS|POSS MARIJ") ~ "Drug",
      str_detect(`OFFENSE-DESC`, "BURGLARY|BURG HAB|BURG VEHICLE") ~ "Burglary",
      str_detect(`OFFENSE-DESC`, "EVADING ARREST") ~ "Evading",
      str_detect(`OFFENSE-DESC`, "MURDER|HOMICIDE|MANSLAUGHTER") ~ "Homicide",
      str_detect(`OFFENSE-DESC`, "AGG ASSLT|ASSLT|INJURY TO CHILD|RETALIATION") ~ "Assault",
      str_detect(`OFFENSE-DESC`, "FORG|CREDIT/DEBIT|FRAUD|THEFT|UNAUTH USE VEH|CRIM MISCH") ~ "Property",
      str_detect(`OFFENSE-DESC`, "DWI|DRIV WHILE INTOX") ~ "DWI",
      str_detect(`OFFENSE-DESC`, "SEX|RAPE|INDECENCY|SEXUAL") ~ "Sex",
      str_detect(`OFFENSE-DESC`, "WEAPON|WPN|CARRY") ~ "Weapon",
      TRUE ~ "Other"
    ),
    OFFENSE_CATEGORY2 = fct_relevel(OFFENSE_CATEGORY2, "Other"),
    SEX          = factor(`SEX-LABEL`),
    APPOINTED    = as.integer(`ATTORNEY-TYPE` == "Appointed"),
    COURT        = factor(COURT),
    CASE_YEAR_FE = factor(`CASE-YEAR`),
    PROSECUTOR   = factor(`INTAKE-PROSECUTOR`),
    BOND_LOG     = log(`BOND-AMOUNT` + 1)
  )

# Drop prosecutors first observed in 1990 (left-censored career counts)
fresh_prosecutors <- df |>
  group_by(`INTAKE-PROSECUTOR`) |>
  summarise(first_year = min(`CASE-YEAR`, na.rm = TRUE), .groups = "drop") |>
  filter(first_year >= 1991) |>
  pull(`INTAKE-PROSECUTOR`)

df <- df |> filter(`INTAKE-PROSECUTOR` %in% fresh_prosecutors)

# Rescale experience to per-100-cases for readable table coefficients
df <- df |> mutate(PROSECUTOR_CASE_N100 = PROSECUTOR_CASE_N / 100)

# Defendant career sequence (proxy for prior record)
df <- df |>
  arrange(SID, `CASE-YEAR`) |>
  group_by(SID) |>
  mutate(DEFENDANT_CASE_N = row_number()) |>
  ungroup()

message("Analysis sample: ", nrow(df), " cases, ",
        n_distinct(df$`INTAKE-PROSECUTOR`), " prosecutors (1991+ only)")

# Helper: run logit, return tidy table
run_logit <- function(formula, data, label) {
  message("\n", strrep("=", 60))
  message("Model: ", label, "  (N = ", nrow(data), ")")
  fit <- tryCatch(
    glm(formula, data = data, family = binomial),
    error = function(e) { message("  ERROR: ", e$message); NULL }
  )
  if (is.null(fit)) return(tibble(model = label))
  tidy(fit, conf.int = TRUE) |> mutate(OR = exp(estimate), model = label)
}

run_feglm <- function(formula, data, fe_vars, label) {
  message("\n", strrep("=", 60))
  message("Model: ", label, "  (N = ", nrow(data), ")")
  fit <- tryCatch(
    feglm(formula, data = data, fixef = fe_vars, family = binomial(),
          fixef.tol = 1e-4, fixef.iter = 50, iter = 50),
    error = function(e) { message("  ERROR: ", e$message); NULL }
  )
  if (is.null(fit)) return(tibble(model = label))
  print(summary(fit))
  tidy(fit, conf.int = TRUE) |> mutate(OR = exp(estimate), model = label)
}

# ── M1: Baseline — no controls ────────────────────────────────────────────────

m1 <- run_logit(
  DEFERRED ~ BLACK + LATINO + PROSECUTOR_CASE_N100 +
    BLACK:PROSECUTOR_CASE_N100 + LATINO:PROSECUTOR_CASE_N100,
  df, "M1_Baseline")

# ── M2: Offense type and category controls ────────────────────────────────────

m2 <- run_logit(
  DEFERRED ~ BLACK + LATINO + PROSECUTOR_CASE_N100 +
    BLACK:PROSECUTOR_CASE_N100 + LATINO:PROSECUTOR_CASE_N100 +
    OFFENSE_TYPE + OFFENSE_CATEGORY2,
  df |> filter(!is.na(OFFENSE_CATEGORY2)),
  "M2_OffenseFE")

# ── Spec A: Clean baseline — offense + attorney controls (main table col 1) ───

specA <- run_logit(
  DEFERRED ~ BLACK + LATINO + PROSECUTOR_CASE_N100 +
    BLACK:PROSECUTOR_CASE_N100 + LATINO:PROSECUTOR_CASE_N100 +
    OFFENSE_TYPE + OFFENSE_CATEGORY2 + APPOINTED,
  df |> filter(!is.na(OFFENSE_CATEGORY2)),
  "SpecA_OffenseOnly")

# ── Spec B: Add case year FE (main table col 2) ───────────────────────────────
# Note: CASE_YEAR_FE is collinear with PROSECUTOR_CASE_N100 by construction;
# interaction surviving here is a conservative test.
# Uses feglm for speed (glm + tidy hangs on 24 year FE dummies).

specB <- run_feglm(
  DEFERRED ~ BLACK + LATINO + PROSECUTOR_CASE_N100 +
    BLACK:PROSECUTOR_CASE_N100 + LATINO:PROSECUTOR_CASE_N100 +
    OFFENSE_TYPE + OFFENSE_CATEGORY2 + APPOINTED,
  df |> filter(!is.na(OFFENSE_CATEGORY2)),
  fe_vars = "CASE_YEAR_FE",
  "SpecB_YearFE")

# ── Spec C: Prosecutor fixed effects (main table col 3) ───────────────────────
# Within-prosecutor identification; answers selection critique.

specC <- run_feglm(
  DEFERRED ~ BLACK + LATINO + PROSECUTOR_CASE_N100 +
    BLACK:PROSECUTOR_CASE_N100 + LATINO:PROSECUTOR_CASE_N100 +
    OFFENSE_TYPE + OFFENSE_CATEGORY2 + APPOINTED,
  df |> filter(!is.na(OFFENSE_CATEGORY2)),
  fe_vars = "PROSECUTOR",
  "SpecC_ProsecutorFE")

# ── M6: Appointed counsel only (robustness) ───────────────────────────────────

m6 <- run_logit(
  DEFERRED ~ BLACK + LATINO + PROSECUTOR_CASE_N100 +
    BLACK:PROSECUTOR_CASE_N100 + LATINO:PROSECUTOR_CASE_N100 +
    OFFENSE_TYPE + OFFENSE_CATEGORY2,
  df |> filter(`ATTORNEY-TYPE` == "Appointed", !is.na(OFFENSE_CATEGORY2)),
  "M6_AppointedOnly")

# ── M7: Within offense type (robustness) ──────────────────────────────────────

m7_list <- map(c("F1", "F2", "F3", "FS"), function(ot) {
  sub <- df |> filter(`OFFENSE-CLASS` == ot, !is.na(OFFENSE_CATEGORY2))
  tryCatch(
    run_logit(
      DEFERRED ~ BLACK + LATINO + PROSECUTOR_CASE_N100 +
        BLACK:PROSECUTOR_CASE_N100 + LATINO:PROSECUTOR_CASE_N100 +
        OFFENSE_CATEGORY + APPOINTED,
      sub, paste0("M7_", ot, "_only")),
    error = function(e) {
      message("  M7_", ot, " skipped: ", e$message)
      tibble(model = paste0("M7_", ot, "_only_SKIPPED"))
    }
  )
})
m7 <- bind_rows(m7_list)

# ── M8: Deferred conditional on any plea (robustness) ────────────────────────

df_plea <- df |> filter(`GUILTY-PLEA` == 1 | DEFERRED == 1, !is.na(OFFENSE_CATEGORY2))

m8 <- run_logit(
  DEFERRED ~ BLACK + LATINO + PROSECUTOR_CASE_N100 +
    BLACK:PROSECUTOR_CASE_N100 + LATINO:PROSECUTOR_CASE_N100 +
    OFFENSE_TYPE + OFFENSE_CATEGORY2 + APPOINTED + CASE_YEAR_FE,
  df_plea,
  "M8_DeferredConditionalOnPlea")

# ── M9: Straight conviction via plea (robustness) ─────────────────────────────

m9 <- run_logit(
  STRAIGHT_CONVICTION ~ BLACK + LATINO + PROSECUTOR_CASE_N100 +
    BLACK:PROSECUTOR_CASE_N100 + LATINO:PROSECUTOR_CASE_N100 +
    OFFENSE_TYPE + OFFENSE_CATEGORY2 + APPOINTED + CASE_YEAR_FE,
  df_plea |> mutate(STRAIGHT_CONVICTION = as.integer(`GUILTY-PLEA` == 1 & DEFERRED == 0)),
  "M9_StraightConviction")

# ── Spec D: Prosecutor + Court crossed FE (addresses judge-learning concern) ──

specD <- run_feglm(
  DEFERRED ~ BLACK + LATINO + PROSECUTOR_CASE_N100 +
    BLACK:PROSECUTOR_CASE_N100 + LATINO:PROSECUTOR_CASE_N100 +
    OFFENSE_TYPE + OFFENSE_CATEGORY2 + APPOINTED,
  df |> filter(!is.na(OFFENSE_CATEGORY2)),
  fe_vars = c("PROSECUTOR", "COURT"),
  "SpecD_ProsecutorCourtFE")

# ── SpecA/C with defendant prior-record proxy (robustness) ───────────────────

specA_priors <- run_logit(
  DEFERRED ~ BLACK + LATINO + PROSECUTOR_CASE_N100 +
    BLACK:PROSECUTOR_CASE_N100 + LATINO:PROSECUTOR_CASE_N100 +
    OFFENSE_TYPE + OFFENSE_CATEGORY2 + APPOINTED + DEFENDANT_CASE_N,
  df |> filter(!is.na(OFFENSE_CATEGORY2)),
  "SpecA_WithPriors")

specC_priors <- run_feglm(
  DEFERRED ~ BLACK + LATINO + PROSECUTOR_CASE_N100 +
    BLACK:PROSECUTOR_CASE_N100 + LATINO:PROSECUTOR_CASE_N100 +
    OFFENSE_TYPE + OFFENSE_CATEGORY2 + APPOINTED + DEFENDANT_CASE_N,
  df |> filter(!is.na(OFFENSE_CATEGORY2)),
  fe_vars = "PROSECUTOR",
  "SpecC_WithPriors")

# ── SpecC controlling for caseload composition drift (addresses selection) ────
# Falsification showed PROSECUTOR_CASE_N100 predicts declining Black/Latino share.
# PRED_BLACK_SHARE is the within-prosecutor model-implied Black probability
# given offense type/category; adding it tests whether the interaction reflects
# caseload sorting rather than differential treatment.

black_share_fit <- feglm(
  BLACK ~ PROSECUTOR_CASE_N100 + OFFENSE_TYPE + OFFENSE_CATEGORY2 | PROSECUTOR,
  data = df |> filter(!is.na(OFFENSE_CATEGORY2)),
  family = binomial()
)

df_comp <- df |>
  filter(!is.na(OFFENSE_CATEGORY2)) |>
  mutate(PRED_BLACK_SHARE = fitted(black_share_fit))

specC_comp <- run_feglm(
  DEFERRED ~ BLACK + LATINO + PROSECUTOR_CASE_N100 +
    BLACK:PROSECUTOR_CASE_N100 + LATINO:PROSECUTOR_CASE_N100 +
    OFFENSE_TYPE + OFFENSE_CATEGORY2 + APPOINTED + DEFENDANT_CASE_N +
    PRED_BLACK_SHARE,
  df_comp,
  fe_vars = "PROSECUTOR",
  "SpecC_CompControl")

# ── Export ────────────────────────────────────────────────────────────────────

all_results <- bind_rows(m1, m2, specA, specB, specC, specD, m6, m7, m8, m9,
                         specA_priors, specC_priors, specC_comp) |>
  select(model, term, estimate, std.error, statistic, p.value, conf.low, conf.high, OR)

write_csv(all_results, file.path(DATA_DIR, "bexar_model_results.csv"))
message("\nSaved: bexar_model_results.csv")

message("\n── KEY COEFFICIENTS: BLACK × PROSECUTOR_CASE_N100 ──")
all_results |>
  filter(str_detect(term, "BLACK.*PROSECUTOR_CASE_N100|PROSECUTOR_CASE_N100.*BLACK")) |>
  select(model, term, estimate, std.error, p.value, OR) |>
  print(n = Inf)

message("\n── KEY COEFFICIENTS: LATINO × PROSECUTOR_CASE_N100 ──")
all_results |>
  filter(str_detect(term, "LATINO.*PROSECUTOR_CASE_N100|PROSECUTOR_CASE_N100.*LATINO")) |>
  select(model, term, estimate, std.error, p.value, OR) |>
  print(n = Inf)

File: bexar_models.py
"""
Bexar County — Regression Models
Requires: bexar_panel_1990_2021.parquet (produced by bexar_clean.py)

Models:
  1. Baseline logistic: DEFERRED ~ RACE + OFFENSE_CLASS + SEX + CASE_YEAR
  2. Add attorney type
  3. Add prosecutor fixed effects (>= 50 cases)
  4. DiD: DEFERRED ~ RACE * POST2019 + controls
  5. Robustness: GUILTY_PLEA as outcome

Output: bexar_model_results.csv (coefficient tables), printed summaries
"""

import warnings
import pandas as pd
import numpy as np
import statsmodels.formula.api as smf
from statsmodels.tools.sm_exceptions import PerfectSeparationWarning

warnings.filterwarnings("ignore", category=PerfectSeparationWarning)

panel = pd.read_parquet("bexar_panel_1990_2021.parquet")

# ── Prep analysis frame ───────────────────────────────────────────────────────

primary = panel[panel["RACE-LABEL"].isin(["Black", "White", "Latino"])].copy()
primary = primary[primary["SEX-LABEL"].isin(["M", "F"])].copy()

# Reference category: White, F3, Hired counsel
primary["RACE_CAT"]     = pd.Categorical(primary["RACE-LABEL"],
                                          categories=["White", "Black", "Latino"])
primary["OFFENSE_CAT"]  = pd.Categorical(
    primary["OFFENSE-CLASS"],
    categories=["F3", "F1", "F2", "FS", "Other Felony", "Misdemeanor"])
primary["SEX_CAT"]      = pd.Categorical(primary["SEX-LABEL"],
                                          categories=["M", "F"])
primary["ATTY_CAT"]     = pd.Categorical(
    primary.loc[primary["ATTORNEY-TYPE"].isin(["Appointed", "Hired"]), "ATTORNEY-TYPE"],
    categories=["Hired", "Appointed"])
primary["POST2019"]     = (primary["CASE-YEAR"] >= 2019).astype(int)
primary["CASE_YEAR_CTR"] = primary["CASE-YEAR"] - 2005  # center for interpretability

# Restrict to known felonies for primary models
felony_mask = primary["OFFENSE-CLASS"].isin(["F1", "F2", "F3", "FS", "Other Felony"])


def run_logit(formula: str, data: pd.DataFrame, label: str):
    print(f"\n{'='*60}")
    print(f"Model: {label}")
    print(f"N = {len(data):,}")
    print(f"{'='*60}")
    try:
        model = smf.logit(formula, data=data).fit(
            method="bfgs", maxiter=500, disp=False)
        print(model.summary2())
        return model
    except Exception as e:
        print(f"  ERROR: {e}")
        return None


# ── Model 1: Baseline ─────────────────────────────────────────────────────────

d1 = primary[felony_mask].dropna(subset=["DEFERRED", "CASE_YEAR_CTR"])
m1 = run_logit(
    "DEFERRED ~ C(RACE_CAT) + C(OFFENSE_CAT) + C(SEX_CAT) + CASE_YEAR_CTR",
    d1, "1 — Baseline (deferred ~ race + offense + sex + year)")

# ── Model 2: Add attorney type ────────────────────────────────────────────────

d2 = primary[felony_mask & primary["ATTORNEY-TYPE"].isin(["Appointed", "Hired"])].copy()
d2 = d2.dropna(subset=["DEFERRED", "CASE_YEAR_CTR"])
m2 = run_logit(
    "DEFERRED ~ C(RACE_CAT) + C(OFFENSE_CAT) + C(SEX_CAT) + C(ATTY_CAT) + CASE_YEAR_CTR",
    d2, "2 — Add attorney type")

# ── Model 3: Prosecutor fixed effects ────────────────────────────────────────

high_vol = primary["INTAKE-PROSECUTOR"].value_counts()
high_vol = high_vol[high_vol >= 50].index
d3 = d2[d2["INTAKE-PROSECUTOR"].isin(high_vol)].copy()
d3["PROSECUTOR_FE"] = d3["INTAKE-PROSECUTOR"].astype("category")
m3 = run_logit(
    "DEFERRED ~ C(RACE_CAT) + C(OFFENSE_CAT) + C(SEX_CAT) + C(ATTY_CAT)"
    " + CASE_YEAR_CTR + C(PROSECUTOR_FE)",
    d3, f"3 — Prosecutor FE (prosecutors with ≥50 cases, N prosecutors = {len(high_vol)})")

# ── Model 4: DiD (race × post-2019) ──────────────────────────────────────────

d4 = d2.copy()
m4 = run_logit(
    "DEFERRED ~ C(RACE_CAT) * POST2019 + C(OFFENSE_CAT) + C(SEX_CAT) + C(ATTY_CAT)",
    d4, "4 — DiD: race × POST2019")

# ── Model 5: Guilty plea (robustness) ────────────────────────────────────────

m5 = run_logit(
    "Q('GUILTY-PLEA') ~ C(RACE_CAT) + C(OFFENSE_CAT) + C(SEX_CAT) + C(ATTY_CAT) + CASE_YEAR_CTR",
    d2, "5 — Robustness: guilty plea as outcome")

# ── Export coefficient table ──────────────────────────────────────────────────

def coef_table(model, label):
    if model is None:
        return pd.DataFrame()
    t = model.summary2().tables[1].copy()
    t.columns = ["Coef", "Std_Err", "z", "P>|z|", "CI_Low", "CI_High"]
    t["OR"] = np.exp(t["Coef"]).round(4)
    t["Model"] = label
    t.index.name = "Variable"
    return t.reset_index()

tables = pd.concat([
    coef_table(m1, "M1_Baseline"),
    coef_table(m2, "M2_AttorneyType"),
    coef_table(m3, "M3_ProsecutorFE"),
    coef_table(m4, "M4_DiD"),
    coef_table(m5, "M5_GuiltyPlea"),
], ignore_index=True)

tables.to_csv("bexar_model_results.csv", index=False)
print("\n\nSaved coefficient tables to: bexar_model_results.csv")

# ── Sentence length OLS (conditional on conviction) ──────────────────────────

convicted = primary[
    felony_mask &
    (primary["CONVICTED"] == 1) &
    (primary["SENTENCE-DAYS"] > 0) &
    primary["ATTORNEY-TYPE"].isin(["Appointed", "Hired"])
].copy()
convicted["LOG_SENTENCE"] = np.log(convicted["SENTENCE-DAYS"])

print(f"\n{'='*60}")
print(f"OLS — Log sentence days | convicted (N = {len(convicted):,})")
print(f"{'='*60}")
try:
    ols = smf.ols(
        "LOG_SENTENCE ~ C(RACE_CAT) + C(OFFENSE_CAT) + C(SEX_CAT) + C(ATTY_CAT) + CASE_YEAR_CTR",
        data=convicted).fit()
    print(ols.summary2())
    ols_tbl = ols.summary2().tables[1].copy()
    ols_tbl["Model"] = "OLS_LogSentence"
    ols_tbl.index.name = "Variable"
    ols_tbl.reset_index().to_csv("bexar_ols_sentence.csv", index=False)
    print("Saved: bexar_ols_sentence.csv")
except Exception as e:
    print(f"  ERROR: {e}")
